mlflow/azureml/__init__.py

```diff
@@ -380,15 +382,45 @@ def deploy(
             _copy_file_or_tree(src=absolute_model_path, dst=model_directory_path),
         )
 
-        registered_model = AzureModel.register(
-            workspace=workspace, model_path=tmp_model_path, model_name=model_name, tags=tags
-        )
+        registered_model = None
+        m_id = None
 
-        _logger.info(
-            "Registered an Azure Model with name: `%s` and version: `%s`",
-            registered_model.name,
-            registered_model.version,
-        )
+        if model_uri.startswith("models:/"):
+            m_name = model_uri.split("/")[-2]
+            m_version = int(model_uri.split("/")[-1])
```

Comment:
The last component of `model_uri` may not always be an integer version. Users can specify the lifecycle stage of a model via the last component. For example, `models:/my-model/production` can be used to refer to the latest version of `my-model` in stage `production`.

Because the latest production version may change over time, `m_id` will not necessarily continue to refer to the same model version. A similar problem occurs if the user deletes a registered model and re-creates it with the same name: version `1` of `my-model` at time T is not guaranteed to refer to the same ML model as version `1` of `my-model` at time T + n if deletion and re-creation occurs in between.

Is `m_id` meant to be a unique identifier? Are we trying to enforce a 1:1 mapping between AzureModels and model versions in the AzureML Model Registry? It would be helpful to spell out the exact goals in the PR description.

Reply:
In AzureML, model ID is unique within a workspace, and is a composite of <name>:<version>. This doesn't necessarily mean that there is a 1:1 mapping between a model file and a model version, as the file could be registered multiple times, deleted and re-registered.etc.

Specifically for this however, we need to have the AzureML Model ID in order to deploy, that's just how the deployment functionality works currently. This function that I'm editing is a workaround that was introduced to allow MLFlow users a convenience method for deploying on Azure, and as such as limitations. I personally am fine with saying that one of those limitations is lack of support for `/production` type uris, however @akshaya-a may feel differently about that.

Reply:
> In AzureML, model ID is unique within a workspace, and is a composite of :. This doesn't necessarily mean that there is a 1:1 mapping between a model file and a model version, as the file could be registered multiple times, deleted and re-registered.etc.

The model ID uniqueness part of this concerns me though. With the logic implemented in this PR, what happens if a user deploys a model with name `A`, deletes the associated registered model, and then tries to redeploy with name `A`? In that case, it seems like we'll pull down the old AzureModel because the ID matches and end up deploying content that the user didn't specify.

Also, do users expect to have their models registered with the MLflow Model Registry on Azure at deployment time? We should at least document this side-effect, but I wonder if that's the intended UX.

Reply:
What are the drawbacks to using a UUID for the model ID instead?

Reply:
A UUID can not be used for ID, the ID is not something that is set client side, it's set service side, and is a composite of name and version as I mention, that is out of my control.

If a user deploys a model with name A version 1, and then deletes that model, it won't exist anywhere at that point, so it couldn't pull down the old model. Unless I'm not understanding your concern correctly?

And yes, it is expected that the model will be registered at deployment time if it's not already. At the very least it is the expectation for our users and has been from the beginning.